cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkUdt.java

```diff
@@ -0,0 +1,172 @@
+           .toArray(StructField[]::new)
+        );
+    }
+
+    @Override
+    public Object sparkSqlRowValue(GenericInternalRow row, int position)
+    {
+        final InternalRow struct = row.getStruct(position, size());
+        return IntStream.range(0, size())
+                        .boxed()
+                        .collect(Collectors.toMap(
+                        index -> field(index).name(),
+                        index -> sparkType(index).toTestRowType(struct.get(index, sparkType(index).dataType())
+                        )));
+    }
+
+    @Override
+    public Object sparkSqlRowValue(Row row, int position)
+    {
+        final Row struct = row.getStruct(position);
+        return IntStream.range(0, struct.size())
+                        .boxed()
+                        .filter(index -> !struct.isNullAt(index))
+                        .collect(Collectors.toMap(
+                        index -> struct.schema().fields()[index].name(),
+                        index -> sparkType(index).toTestRowType(struct.get(index))
+                        ));
+    }
+
+    public List<SparkType> sparkTypes()
+    {
+        return sparkTypes;
+    }
+
+    public int size()
+    {
+        return udt.fields().size();
+    }
+
+    public SparkType sparkType()
+    {
+        return sparkType(0);
+    }
+
+    public SparkType sparkType(int position)
+    {
+        return sparkTypes().get(position);
+    }
+
+    public int compareTo(Object first, Object second)
+    {
+        return SparkType.compareArrays(((GenericInternalRow) first).values(), ((GenericInternalRow) second).values(), this::sparkType);
+    }
+
+    @Override
+    public boolean equals(Object first, Object second)
+    {
+        return SparkType.equalsArrays(((GenericInternalRow) first).values(), ((GenericInternalRow) second).values(), this::sparkType);
+    }
+
+    @Override
+    public Object toTestRowType(Object value)
+    {
+        GenericRowWithSchema row = (GenericRowWithSchema) value;
+        String[] fieldNames = row.schema().fieldNames();
+        java.util.Map<String, Object> result = new LinkedHashMap<>(fieldNames.length);
+        for (int index = 0; index < fieldNames.length; index++)
+        {
+            result.put(fieldNames[index], sparkType(index).toTestRowType(row.get(index)));
+        }
+        return result;
+    }
+
+    @Override
+    public Object toSparkSqlType(Object value, boolean isFrozen)
+    {
+        return udtToSparkSqlType(value, isFrozen);
+    }
+
+    @SuppressWarnings("unchecked")
+    private GenericInternalRow udtToSparkSqlType(Object value, boolean isFrozen)
+    {
+        if (value instanceof ByteBuffer)
+        {
+            // Need to deserialize first, e.g. if UDT is frozen inside collections
+            return udtToSparkSqlType(udt.deserializeUdt(converter, (ByteBuffer) value, isFrozen));
+        }
+        else
+        {
+            return udtToSparkSqlType((java.util.Map<String, Object>) value);
```

Comment:
nit: add a comment to explain where the `Map<String, Object>` is coming from. (org.apache.cassandra.spark.data.complex.CqlUdt#deserializeUdt)

I do not think the full qualified class name is required for map.